### ISSUE TITLE
Change maven group to io.jaegertracing

### ISSUE THREAD

**Issue by objectiser** [jaegertracing/jaeger-client-java#287]
As discussed on friday's call, the maven group should be changed from `com.uber.jaeger` to `io.jaegertracing`. The proposal was to make this change in the `v0.22.0` branch so that it gets merged to master as part of the OT Java 0.31.0 change. 

However the implication is that the Java package would also change (as it usually follows the maven group), and this would be a breaking change - so wanted to make sure this is clear (as wasn't explicitly discussed)?

Other option would be to make this change when moving to version 1,0.

**Comment by objectiser** [comment]
This is probably the last change we should do prior to releasing the 1.0 version, just in case we want to do other interim releases. So knowing that the build infrastructure can cope is good - but we might want to hold off for a bit.

**Comment by yurishkuro** [comment]
Agreed. We just had a breaking/disruptive change, many users at Uber are already grumpy about it. Change in the package name has no functional benefits, just another disruption.

**Comment by pavolloffay** [comment]
@yurishkuro do you want to still publish this artifact under `com.uber`? Is it required?

If so, there might be these solutions:

## Plan A
1. change groupid to `io.jaegertracing`
2. configure gradle publish process to publish under `io.jaegertracing` and also `com.uber`

With this change we would like to tackle also #182:
1. move instrumentations into a separate repository - preferably to uber github organization. If they cannot be moved we could keep them here and just don't publish then under `io.jaegertracing`

## Plan B
1. rename to `io.jaegertracing`
2. create wrappers for `com.uber` core classes .e.g. tracer, configuration but also senders, reporters, sampler. Probably all classes? :/ 
3. move instrumentations into a separate repo

**Comment by yurishkuro** [comment]
I like the idea of publishing two sets of artifacts, but not sure how messy it would make the build. How about this:

- clone jaeger-core into jaeger-core-new module
- update new module to io.* packages, do a clean up/refactoring as we see fit
- change core module to be a wrapper or subclasses of core-new, and deprecate everything
- make sure that all original tests pass in both cores, to confirm we didn't break the compatibility

Once we reach this point, we can split the repos, leaving only core-new, rename it to core and change the artifact namespace.

**Comment by pavolloffay** [comment]
If we add a new class to new core, do you want to add it back to com.uber and deprecate it right-away?

**Comment by yurishkuro** [comment]
Don't think so - what kind of class do you have in mind?

**Comment by pavolloffay** [comment]
For instance new sender, sampler. If we don't add it back, why do we need to keep `com.uber` ?

**Comment by yurishkuro** [comment]
We only want to keep uber.com for backwards compatibility. Adding new types has no backwards compatibility requirements.

**Comment by yurishkuro** [comment]
PS: we really need to see if doing a wrapper set of `uber.com.*` classes is even feasible...

**Comment by pavolloffay** [comment]
I will have a look at it shortly.

**Comment by pavolloffay** [comment]
## Wrapper approach steps
1. create new repo `jaeger-client-java-new` for io.jaegertracing
2. copy `core, micrometer, xdock, zipkin, thrift` modules to the new repo
3. make old repo depend on io.jaegertracing modules and create extends classes for io.jaegertracing.

What does it break?
* package local symbols will not be accessible. E.g. 3rd party dep created package com.uber and used package local symbols
* if we rename/remove class in io.jagertracing it will break - changes in com.uber will be required.
* if a method/symbol is changed/removed it's removed from `com.uber`. Although these two steps should be fine as we will keep backward compatibility in io.jaegertracing

We will continue development in io.jaegertracing which can be versioned independently of the old repo. Old repo will be versioned and released separately.

I assume that after some time we could stop supporting the old com.uber.

## Thrift java problems
There is a problem with thrift classes. thrift defines package `com.uber`. To solve it we could use the same approach - add thrift classes to the repo which will extend io.jaegertracing.

**Comment by objectiser** [comment]
@pavolloffay Don't forget the `tracerresolver`module copy to the new repo.

**Comment by pavolloffay** [comment]
yes, hmm I think it will be better to use just one repo with two branches? WDYT? Master for io.jaegertracing and com.uber for the old one.

**Comment by objectiser** [comment]
Not sure - thought that the `com.uber` version/repo would be better moved into Uber's github org.

**Comment by pavolloffay** [comment]
Extend a class approach does not work: if a com.uber class extends io.jaegertracing then all methods in the extended class uses symbols from io.jaegertracing - which is not what we want.

Wrapper approach doesn't work either. For example `BaggageSetter.setBaggage(span, k, v)` https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/com/uber/jaeger/baggage/BaggageSetter.java#L52 gets com.uber.jaeger.Span which cannot be passed to wrapped io.jaegertracing `BaggageSetter` because it consumes Span from io. The span has a package private constructor.

**Comment by pavolloffay** [comment]
I and @jpkrohling could make it work - both extend and wrapper approach. 

I can think of two last approaches
1. Just maintain branch with com.uber for bug fixes and also add deprecated to all classes.
2. use shade plugin to create com.uber - so the release will publish two artifacts with different groupids

@yurishkuro WDYT?

**Comment by jpkrohling** [comment]
> could make it work

Could *not* make it work ;)

**Comment by yurishkuro** [comment]
> use shade plugin to create com.uber

I was also thinking about this, seems like a better approach than a fork.

**Comment by pavolloffay** [comment]
Based on the discussion on bi-weekly meeting we are going to keep to codebases. This repo will hold `io.jaegertracing` and fork `com.uber.jaeger`.
